README.md
<h1 align="center">
  Scrapix
</h1>
<h4 align="center">
    A TypeScript-based CLI tool for scraping and downloading Google images 🚀.
</h4>

<div align="center">
  <img src="https://badgen.net/npm/v/make-cli-tool?icon=npm" />
  <img src="https://badgen.net/github/last-commit/sajmoni/make-cli-tool?icon=github" />
</div>

[![experimental](http://badges.github.io/stability-badges/dist/experimental.svg)](http://github.com/badges/stability-badges)

### Description
`Scrapix` is an interactive CLI for scraping and downloading Google images easily. The core of this CLI is the [images-scraper](https://www.npmjs.com/package/images-scraper) package, which uses a puppeteer-based headless browser to automatically search and return the URLs of the images. This `scrapix` CLI can help developers and non-developers to build image datasets for computer vision tasks i.e. image classification, object detection, face recognition. 

### Installation
```bash
$ npm install -g scrapix-cli
```

### To resolve `puppeteer` installation issues on Linux based OS: -
```bash
$ sudo npm install -g scrapix-cli --unsafe-perm=true
```

### Usage 
To start the CLI, Type the `scrapix` command in your CMD or terminal: -
```bash
$ scrapix
```
### Modes 
| default | The search keywords are defined in the terminal directly |
|---------|----------------------------------------------------------|
| file    | The search keywords are defined in a .json file.         |

##### Default mode: usage
Required Parameters: -
- `keywords`: - The names of images you want to download seperated by commas (i.e. roses,daisy,hibiscus)
- `number`: - The number of images you want to download

##### File mode: usage
Required parameters: -
- Name of .json file in the base directory to load the keyword and number of images to be scraped and downloaded for each keyword.

##### Structure of the .json file content
```json
{
    "images": [
        {"keyword": "roses", "number":20},
        {"keyword":"daisy", "number":25},
        {"keyword":"hibiscus", "number":30},
    ]
}
```

[View Video Demonstration](https://www.youtube.com/watch?v=3mRXiu-i4TY)

### TODOS
- Image validation 
- Support for custom image processing actions i.e. resizing, compression
- Improve error handling
- Provide support for other search engines i.e. Bing, Wikipedia
- Improve general performance
- Clean code and develop a better documentation

### Contributing Guide
Coming soon 


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.1.8](https://github.com/rexsimiloluwah/scrapix-cli/compare/v0.1.7...v0.1.8) (2021-10-14)

### [0.1.7](https://github.com/rexsimiloluwah/scrapix-cli/compare/v0.1.6...v0.1.7) (2021-08-17)

### [0.1.6](https://github.com/rexsimiloluwah/scrapix-cli/compare/v0.1.5...v0.1.6) (2021-08-16)

### [0.1.5](https://github.com/rexsimiloluwah/scrapix-cli/compare/v1.0.1...v0.1.5) (2021-08-16)

### 0.1.4 (2021-08-15)


src/index.ts
#!/usr/bin/env node

import { program } from 'commander';
import { readFileSync } from 'fs';
import path from 'path';
import clear from 'clear';
import chalk from 'chalk';
import figlet from 'figlet';
import { performance } from 'perf_hooks';
import { Inquirer } from './utils/inquirer';
import { scrapeImages, FileContent } from './utils/scraper';


const { description, version } = require('../package.json');

if (!process.argv.slice(2).length) {
    clear() // To clear the terminal
    console.log(
        chalk.red(
            figlet.textSync('Scrapix', { horizontalLayout: 'full' })
        )
    )
    console.log(chalk.gray('⚡ A CLI tool for scraping Google images the right way'))
}

program
    .description(description)
    .version(chalk.green(`Version: ${version}`), '-v')
    .parse(process.argv);

// start the CLI 
function measurePromise(fn: () => Promise<any>): Promise<number> {
    let onPromiseDone = () => performance.now() - start;

    let start = performance.now();
    return fn().then(onPromiseDone, onPromiseDone);
}

const start = async () => {
    await Inquirer().then(results => {
        const { input, number, filename } = results;
        if(filename){
            const file_contents = JSON.parse(readFileSync(path.join('.', filename), 'utf8'));
            if(!file_contents['images'] || !file_contents['images'].length){
                console.log("File must contain the required images.");
            }else{
                file_contents['images'].map((item:FileContent, id:number)=>(
                    scrapeImages(item['keyword'].trim(), item['number'])
                ))
            }
        }
        else{
            input.split(',').map((keyword:string)=>(
                scrapeImages(keyword.trim(), number)
            ))
        }
    }).catch(error=>{
        console.log(`An unknown error occurred --> ${error}`)
    })
}

start();


src/utils/scraper.ts
/**
 * @author [Similoluwa Okunowo (The Caveman)]
 * @email [rexsimiloluwa@gmail.com]
 * @create date 2020-12-16 20:13:17
 * @modify date 2020-12-21 08:14:27
 * @desc [description]
 */

import fs from 'fs';
import path from 'path';
import request from 'request';
import Scraper from 'images-scraper';

// Initialize the scraper 
const google = new Scraper({
    puppeteer: {
        headless: false,
    }
})

export interface ScrapedImage {
    id: number,
    url: string,
    source?: string,
    title?: string
}

export interface FileContent {
    keyword: string, 
    number: number
}

// Helper functions 

const file_headers: object = {
    'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8',
    'Accept-Encoding': 'gzip, deflate, br',
    'Accept-Language': 'en-US,en;q=0.9,fr;q=0.8,ro;q=0.7,ru;q=0.6,la;q=0.5,pt;q=0.4,de;q=0.3',
    'Cache-Control': 'max-age=0',
    'Connection': 'keep-alive',
    'Upgrade-Insecure-Requests': '1',
    'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/68.0.3440.106 Safari/537.36'
}


export async function saveFileFromURL(url: string, destination: string): Promise<void> {
    return new Promise<void>((resolve, reject) => {
        request.get(url, file_headers)
            .on('response', (response) => {
                console.log(response.statusCode)
            })
            .on('error', (err) => {
                fs.unlinkSync(destination);
                reject(`${destination} --> ${err}`);
            })
            .pipe(fs.createWriteStream(destination))
            .on('close', () => {
                resolve();
            })

    })
}

export async function saveScrapedImages(image: ScrapedImage, name: string): Promise<void> {
    if (!fs.existsSync(path.join('.','scraped-images'))){ fs.mkdirSync(path.join('.', 'scraped-images')) };
    if (!fs.existsSync(path.join('.',`scraped-images/${name}`))) {
        fs.mkdir(path.join('.', `scraped-images/${name}`), { recursive: true }, (err) => {
            console.log(err)
        })
    }
    //console.log(`Directory ${name} created successfully !`)
    saveFileFromURL(image.url, path.join('.', `scraped-images/${name}/${image.id}.jpg`))
        .then(()=>{
            console.log('Successfully downloaded file.');
        }).catch(error=>{
            console.log(`An error occurred: - ${error}.`);
        })
}

export async function scrapeImages(keyword: string, num: number): Promise<void> {
    await google.scrape(keyword.toLowerCase().trim(), num)
        .then(results => {
            let imageUrls = Object.values(results).map((item: ScrapedImage, id: number) => (
                { id: id, url: item.url }
            ))

            for (let idx in imageUrls) {
                console.log(imageUrls[idx])
                saveScrapedImages(imageUrls[idx], keyword)
            }
        })
        .catch(error => {
            console.log(`An error occurred: - ${error}.`)
        })

}



src/utils/inquirer.ts
/**
 * @author [Similoluwa Okunowo (The Caveman)]
 * @email [rexsimiloluwa@gmail.com]
 * @create date 2020-12-16 20:13:17
 * @modify date 2020-12-21 08:14:27
 * @desc [description]
 */
import path from 'path';
import {readFileSync, existsSync} from 'fs';
import inquirer, { Question, ChoiceOptions, Answers, QuestionCollection } from 'inquirer';

export interface Choice {
    name: string,
    value: Mode | number
}

export enum Mode {
    FILE = 'file',
    DEFAULT = 'default'
}

export async function Inquirer(): Promise<Answers> {
    const modes: Array<ChoiceOptions> = [
        { name: 'default', value: 'default' },
        { name: 'file', value: 'file' }
    ];

    const prompts: QuestionCollection<Answers> = [
        {
            name: 'mode',
            type: 'list',
            message: 'Select mode to load search queries: - ',
            choices: modes
        },
        {
            name: 'input',
            type: 'input',
            message: 'Enter the names of images you want to download i.e. Cat,Dog: - ',
            validate: async (input: string) => {
                if (input.length) { return true }
                return "Empty or Invalid input, seperate the names with commas."
            },
            when: (answers) => answers.mode == 'default'
        },
        {
            name: 'number',
            type: 'input',
            message: 'Enter the number of images to scrape: -',
            validate: async (input: number) => {
                if(!Number(input) || input<0) {
                    return "Input must be a number and greater than 0"
                }
                return true
            },
            when: (answers) => answers.mode == 'default'
        },
        {
            name: 'filename',
            type: 'input',
            message: 'Enter the name of the .json file: -',
            validate: async (input: string) => {
                if(!input.length){ return "Invalid filename, filename cannot be empty." }
                if(path.extname(input)!='.json'){ return "File must be a .json file." }
                if(!existsSync(path.join('.', input))){ return "File does not exist in the base path." }
                if(!readFileSync(path.join('.', input))){ return "File is empty, Kindly follow the specified format." }
                return true
            },
            when: (answers) => answers.mode == 'file'
        }
    ]
    return inquirer.prompt(prompts);
}